subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1662 @@
+		wlk_addr = fs->ate_wra;
+#endif
+		/* initialize the previous_sector_num to an impossible value
+		 * to read the sector cycle in the first loop
+		 */
+		previous_sector_num = -1;
+
+		do {
+			wlk_prev_addr = wlk_addr;
+			rc = zms_prev_ate(fs, &wlk_addr, &wlk_ate);
+			if (rc) {
+				return rc;
+			}
+			/* if ate with same id is reached we might need to copy.
+			 * only consider valid wlk_ate's. Something wrong might
+			 * have been written that has the same ate but is
+			 * invalid, don't consider these as a match.
+			 */
+			if (wlk_ate.id == gc_ate.id) {
+				/* read the ate cycle only when we change the sector
+				 * or if it is the first read
+				 */
+				rc = zms_get_cycle_on_sector_change(
+					fs, wlk_prev_addr, previous_sector_num, &current_cycle);
+				if (rc) {
+					return rc;
+				}
+				if (zms_ate_valid_different_sector(fs, &wlk_ate, current_cycle)) {
+					break;
+				}
+				previous_sector_num = SECTOR_NUM(wlk_prev_addr);
+			}
+		} while (wlk_addr != fs->ate_wra);
+
+		/* if walk has reached the same address as gc_addr copy is
+		 * needed unless it is a deleted item.
+		 */
+		if ((wlk_prev_addr == gc_prev_addr) && gc_ate.len) {
+			/* copy needed */
+			LOG_DBG("Moving %d, len %d", gc_ate.id, gc_ate.len);
+
+			if (gc_ate.len > 8) {
+				/* Copy Data only when len > 8
+				 * Otherwise, Data is already inside ATE
+				 */
+				data_addr = (gc_prev_addr & ADDR_SECT_MASK);
+				data_addr += gc_ate.offset;
+				gc_ate.offset = (uint16_t)(fs->data_wra & ADDR_OFFS_MASK);
+
+				rc = zms_flash_block_move(fs, data_addr, gc_ate.len);
+				if (rc) {
+					return rc;
+				}
+			}
+
+			gc_ate.cycle_cnt = previous_cycle;
+			zms_ate_crc8_update(&gc_ate);
+			rc = zms_flash_ate_wrt(fs, &gc_ate);
+			if (rc) {
+				return rc;
+			}
+		}
+	} while (gc_prev_addr != stop_addr);
+
+gc_done:
+
+	/* restore the previous sector_cycle */
+	fs->sector_cycle = previous_cycle;
+
+	/* Make it possible to detect that gc has finished by writing a
+	 * gc done ate to the sector. In the field we might have zms systems
+	 * that do not have sufficient space to add this ate, so for these
+	 * situations avoid adding the gc done ate.
+	 */
+
+	if (fs->ate_wra >= (fs->data_wra + ate_size)) {
+		rc = zms_add_gc_done_ate(fs);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	/* Erase the GC'ed sector when needed */
+	if (ebw_required) {
+		rc = zms_flash_erase_sector(fs, sec_addr);
+		if (rc) {
+			return rc;
+		}
+	}
+	/* Adding empty ate to the gc'ed sector increments fs->ate_wra
+	 * Restore the ate_wra of the current sector after this
+	 */
+	previous_ate_wra = fs->ate_wra;
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+	zms_lookup_cache_invalidate(fs, sec_addr >> ADDR_SECT_SHIFT);
+#endif
+	rc = zms_add_empty_ate(fs, sec_addr);
+	fs->ate_wra = previous_ate_wra;
+
+	return rc;
+}
+
+int zms_clear(struct zms_fs *fs)
+{
+	int rc;
+	uint64_t addr;
+	bool ebw_required =
+		flash_params_get_erase_cap(fs->flash_parameters) & FLASH_ERASE_C_EXPLICIT;
+
+	if (!fs->ready) {
```

Comment:
Shouldn't here be the mutex check? The function is supposed to be called on mounted system?

Reply:
Yes we should lock the mutex and unlock it after finishing the clear.